### PR TITLE
ARROW-7526: [C++][Compute] Optimize small integer sorting

### PR DESCRIPTION
**Pull request by cyb70289**
Current sorting kernel handles all data types with stl stable_sort.
It is suboptimal for small integers with small value range.

This patch leverages counting sort for Int8 and UInt8 sorting. Up to
10x performance improvement is observed on x86, and above 20x on Arm.

Benchmark
=========
- Tested on Intel E5-2650 and Arm64 CPU
- Data collected by running "arrow-compute-sort-to-indices-benchmark"
- No performance regression is found for Int64 benchmark
- Int8 benchmark is added, result shows big performance boost

Int8 sorting benchmark on E5-2650
---------------------------------
| current    | this patch  | uplift | test case                     |
| -------    | ----------  | ------ | ---------                     |
| 10.0461M/s | 75.2091M/s  | 7.5x   | null_percent=0  size=32.768k  |
| 9.99244M/s | 75.0697M/s  | 7.5x   | null_percent=1  size=32.768k  |
| 10.8323M/s | 71.1796M/s  | 6.6x   | null_percent=10 size=32.768k  |
| 16.9689M/s | 58.756M/s   | 3.5x   | null_percent=50 size=32.768k  |
| 8.25916M/s | 74.9076M/s  | 9.1x   | null_percent=1  size=1048.58k |
| 6.8773M/s  | 73.1168M/s  | 10.6x  | null_percent=1  size=8.38861M |
| 11.9212M/s | 57.5259M/s  | 4.8x   | null_percent=50 size=8.38861M |
| 28.6507M/s | 109.871M/s  | 3.8x   | null_percent=80 size=8.38861M |
| 119.498M/s | 192.564M/s  | 1.6x   | null_percent=99 size=8.38861M |

One interesting finding is current code performs better with more null
values, while this patch performs worst at 50% null values. It's
because current code partitions null values and sort only valid ones,
while this patch treats null as largest value and sort as normal ones.
Branch prediction for null testing works badly if half values are null.
For the extreme case with 99% null values, this patch is 1.6x faster
than current code.

NOTES
=====
Counting sort is not friendly to short arrays. I benchmarked very short
Int8 arrays. Though the test result is very noisy, it does show a bit
performance drops for arrays with fewer than 32 elements on my test bed.

| bytes | current | this patch | uplift |
| ----- | ------- | ---------- | ------ |
| 16    | 16.5M/s | 15.4M/s    | -6%    |
| 32    | 25.4M/s | 25.3M/s    |  0%    |
| 64    | 32.5M/s | 37.9M/s    | +17%   |

It's possible to fallback to stl stable_sort when sorting very short
Int8 arrays. I think it's not necessary as the absolute sorting time
is too short to influence a real world application.

### CODE REVIEW

File: cpp/src/arrow/compute/kernels/sort_to_indices.cc [hunk 1]
```diff
@@ -107,37 +176,34 @@ class SortToIndicesKernelImpl : public SortToIndicesKernel {
...
+template <typename ArrowType, typename Comparator,
+          typename Sorter = CompareSorter<ArrowType, Comparator>>
+SortToIndicesKernelImpl<ArrowType, Sorter>* MakeSortToIndicesKernelImpl(
     Comparator comparator) {
-  return new SortToIndicesKernelImpl<ArrowType, Comparator>(comparator);
+  return new SortToIndicesKernelImpl<ArrowType, Sorter>(Sorter(comparator));
+}
+
+template <typename ArrowType, typename Sorter = CountSorter<ArrowType>>
+SortToIndicesKernelImpl<ArrowType, Sorter>* MakeSortToIndicesKernelImpl(int min,
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/sort_to_indices.cc]: For clarity, perhaps give different names to those two overloads? For example `MakeSortToIndicesWithComparator` and `MakeSortToIndicesCounting`?

**Review by pitrou — APPROVED** [review]
+1, this is a nice improvement. Just a nit.

File: cpp/src/arrow/compute/kernels/sort_to_indices.cc (hunk 1 above)
Comment (cyb70289) [reply cpp/src/arrow/compute/kernels/sort_to_indices.cc]: done

**Comment by pitrou** [comment]
Thank you @cyb70289 !

**Comment by wesm** [comment]
Thanks! Note that this approach also yields perf improvements for "small range" sorts on larger integer types, even when you add in the cost of computing the range

see https://issues.apache.org/jira/browse/ARROW-1571

Perhaps someone will explore this work in the future

**Pull request closed without merging**